### ISSUE TITLE
Breaking change between version 1.3 and 2 in the expected timeseries resultset?

### ISSUE THREAD

**Issue by erikbos** [grafana/azure-data-explorer-datasource#33]
I just deployed version 2 of the plugin on a testbox, and noticed metric names are now constructed in a different way compared to version v1.3 of the plugin.

plugin v1.3:

![Screenshot 2019-09-25 at 14 41 32](https://user-images.githubusercontent.com/1421163/65601651-b839c300-dfa2-11e9-89ce-1dbd4743cfc9.png)

plugin v2.0:

![Screenshot 2019-09-25 at 14 40 39](https://user-images.githubusercontent.com/1421163/65601634-b2dc7880-dfa2-11e9-96da-8b9ac0af9859.png)

It appears that:
- metric name and a metric name's label are swapped around.
- metric now gets a label which is the database column containing the metric values.

(I am on the fence at the moment if labelling metrics adds value here.. they would all get the same label..)

--
*Steps to reproduce*
- Add Graph panel
- Add metric with query below
- Format as Timeseries
- Select time window '2019-09-24 00:00:00' to '2019-09-24T00:30:00'

--
let events = datatable(timestamp:datetime, metricname:string, value:long)
[
    '2019-09-24T00:00:00Z','metric5',47815,
    '2019-09-24T00:00:00Z','metric3',9009,
    '2019-09-24T00:00:00Z','metric2',227900,
    '2019-09-24T00:00:00Z','metric4',40200,
    '2019-09-24T00:00:00Z','metric1',536720,
    '2019-09-24T00:10:00Z','metric3',9908,
    '2019-09-24T00:10:00Z','metric2',227792,
    '2019-09-24T00:10:00Z','metric5',56150,
    '2019-09-24T00:10:00Z','metric4',42821,
    '2019-09-24T00:10:00Z','metric1',539793,
    '2019-09-24T00:20:00Z','metric4',51409,
    '2019-09-24T00:20:00Z','metric1',520867,
    '2019-09-24T00:20:00Z','metric2',243148,
    '2019-09-24T00:20:00Z','metric5',48020,
    '2019-09-24T00:20:00Z','metric3',10664,
    '2019-09-24T00:30:00Z','metric1',530727,
    '2019-09-24T00:30:00Z','metric2',238394,
    '2019-09-24T00:30:00Z','metric5',35661,
    '2019-09-24T00:30:00Z','metric3',10000,
    '2019-09-24T00:30:00Z','metric4',44866,
];
events

**Comment by erikbos** [comment]
I found the issue, the expected query resultset has changed between v1.3 and v2.0:

From https://github.com/grafana/azure-data-explorer-datasource/commit/e4fa6d8472ed0183b30c53162891f19828f5125e#diff-04c6e90faac2675aa89e2176d2eec7d8#L163

So v1.3 expects:

datetime column, a metric name column and a value column

v2.0 expects:

datetime column, named metric 1, named metric2, etc.

I actually like (and want :) !) the functionality as this delivers https://github.com/grafana/azure-data-explorer-datasource/issues/9 I just didn't expect the old queries to break.

**Comment by gilkgit** [comment]
I'm also having issues with that labeling technique.
let's assume this Kusto query:
datatable ( Host: string,Datacenter: string, Count: int)["H1", "DC1",100, "H2", "DC2",150]

So, I tried to achieve drill down ability either using Graph Panel or Bar Gauge (Grafana 6.4 beta 2, Azure Data Explorer plugin V2).

For example in Bar Gauge, when I formatted query results as table
I wanted the labels presented to the user to show H1,  DC1, and the URL to have something similar to var-host=H1&var-datacenter=DC1

I couldn't get to anything useful, not even remotely.
${__series.name} shows constant 'A', which is the label given for that Query in the Query editor

![image](https://user-images.githubusercontent.com/55378597/65631274-e2f63c80-dfdf-11e9-92c1-0976d75fc604.png)

I can get around it by using $__cell_{n} for the field title, but it does not work for the Data Links title
I also tried to add time column and format as Time series on bar gauge and on chart.
There some things work better but others are worse.
Tried many combinations but nothing useful for this drill down scenario.

mind that https://grafana.com/docs/features/panels/graph/ states that charts support the following syntax:
__series.labels.<LABEL> - label’s value to the URL. If your label contains dots use __series.labels["<LABEL>"] 

but this did nothing for Azure Data Explorer datasource plugin

By-the-way, Graph panel visualization with mode "Series" when query is formatted as table (and has a datetime column), also shows only one series named "A", ignoring all data

**Comment by erikbos** [comment]
hi @gilkgit this issue is about a (breaking) change between version 1.3 and 2 in the expected timeseries resultset from the database. One way it manifests itself is by label metrics in a particular way.

It's not about passing parameters via URL. You might want to open a separate for that..

**Comment by srclosson** [comment]
@erikbos, I'm looking at this one. Can you tell me what the original expected behavior was? For example:

This is a valid case:
`let events = datatable(timestamp:datetime, metricname:string, value:long)`

But is this a valid case?
`let events = datatable(timestamp:datetime, metricname:string, value:long, metricname2:string, value2:long)`

How about this one?
`let events = datatable(timestamp:datetime, heartrate:float, systolic:long, diastolic:long)`

And what about tagging/metadata?
`let events = datatable(timestamp:datetime, metricname:string, value:long, datacenter:string, region:string)`

Thoughts?

**Comment by erikbos** [comment]
I think the previous version supported only the first flavour from README.md, L163

> *Time Series* queries are for the Graph Panel (and other panels like the Single Stat panel) and must contain a datetime column, a metric name column and a value column.

https://github.com/grafana/azure-data-explorer-datasource/commit/e4fa6d8472ed0183b30c53162891f19828f5125e#diff-04c6e90faac2675aa89e2176d2eec7d8#L163 

but I have not checked what the code actually allowed/supported.

**Comment by erikbos** [comment]
Perhaps another angle, I noticed we have queries in a panels producing:
`let events = datatable(value:long, metricname:string, timestamp:datetime)

depending on how you read the v1.3 docs you could consider this ok, or in reverse order :) (I think it's reversed and hence we did it not properly)

For me us I am fine with v2's change as the "a column per metric" approach is much more explicit & deterministic. As long this stays we'll just update and simplify the queries to this resultset output when deploying v2.

The questions is perhaps how far you want to go with implementing backwards compatibility?

**Comment by srclosson** [comment]
Personally, I like the approach of doing as much in the query you can, and an aliasing field. For example:

`data | project time=timestamp, metric=metricname, value`

If the columns required for a timeseries query were hardcoded like above (requires time, metric, and value, or for backwards compatibility: timestamp, metricname, and value. The advantage of hardcoding the column names is that ordering of the columns would not be important. 

Then, hardcode a "tags" column. Have as many as you want. Tags would be indexed by the column name:

`data | project time=timestamp, metric=metricname, value, tags=datacenter, tags=region`

Then in the alias, if you used
`metric|tag_datacenter|tag_region`

You might get a naming like:
metric5|Seattle|US-NorthWest 

as your name...

**Comment by srclosson** [comment]
@erikbos: Do you maybe want to try out this and see if it works for you?

**Comment by gilkgit** [comment]
> hi @gilkgit this issue is about a (breaking) change between version 1.3 and 2 in the expected timeseries resultset from the database. One way it manifests itself is by label metrics in a particular way.
> 
> It's not about passing parameters via URL. You might want to open a separate for that..

I didn't write the code but It seems that what you call "metric name" is the thing that is emitted to the URL when using ${__series.name}, so possibly one effects the other.
So, since experiences are probably tangled, I would want devs to consider the URL variables experience when they change the presentation experience.

**Comment by erikbos** [comment]
> @erikbos: Do you maybe want to try out this and see if it works for you?

But there is no "Alias By" text box present...

The templateable approach is maybe a bit feature rich: I prefer to "control" what metrics are visualised in one location by just writing the right Kusto query. I rather not also have to configure/map columns once more out the Kusto query to get the visualisation to work.

**Comment by erikbos** [comment]
About the two potential formats of a Kusto metric(s) resultset, it's not about (personal) preference: I think there are just multiple use cases:

1 `data | project time=timestamp, metric=metricname, value`

This one works well where the source has an metric-per-row design, this also allows metrics to be optional. Doing this way also appears to have the side effect that Grafana starts interpolating when not all metrics are present. (Not what I want)

and

2 `data | project time=timestamp, metric1, metric2`

Works well for situation where each metric must be present for each timestamp. We have noticed that for some tables / queries it's really hard to write a query that produces format #1 when you need all metrics per timestamp in the result set. (We solved it by adding multiple queries to a single panel)

**Comment by srclosson** [comment]
Hey @erikbos, @gilkgit 

Yes, it is more than you asked, but a much more flexible solution all together. It will allow custom tagging. The backend datasource needs to be rebuilt, and restarted for this to work. I you need help with that, I can assist. Just let me know and I can video call or whatever works for you.

The change won't break url variables (it's separate from that, nor will it break any of the $__cell referencing. It was an idea borrowed from the influx datasource, allowing you to custom build your own tag name convention.

The only hardcoded template is $value. This represents any value, so if your value columns were "cpuload" and "memusage", referencing $value will substitute in either "cpuload" or "memusage"

So the request

`data | project timestamp, cpuload, memusage, systemname`

with the alias

`$systemname.$value`

Will yield names like "server1.memusage" and "server2.cpuload"

`cpudata | project timestamp, metric, value, systemname`

Where metric could be things like "cpuuser" and "cpusystem", the alias:

`$metric`: Will yield names like "server1"

`$metric.cpuusage`: Will show server1.cpuusage
`$metric.$value`: Will show "server1.value"

`cpudata | project timestamp server=metric, cpuuser=value
`$metric.$value`: Will show "server1.cpuuser"

Hope this makes sense. Give it a try and let me know if I can help.

**Comment by mpwillia** [comment]
This is a frustrating change for my team. Our issue here is that the change assumes we have a static known set of metrics that we can alias by. In our case we don't. We know what the values _can_ be, but we don't necessarily know when or if they'll show up within any given time period. If they don't show up, we don't want to plot them.

As @erikbos said, the format of `data | project time=timestamp, metric=metricname, value` allows not all metrics to be present for every timestamp. This implicitly makes queries written this way extremely flexible. For example, if we added a new metric type we wouldn't need to modify our Grafana dashboard at all, the query already handles it.

I also agree with @erikbos that there are two distinct use cases here. Perhaps a better solution is to allow the user to select their preference? They can choose to have their time series interpreted as either
`data | project time=timestamp, metric=metricname, value`
or
`data | project time=timestamp, metric1, metric2`

Alternatively, allowing the user control over the series labeling format would also resolve this issue for at least my team. For example, if we could define whether or not to use the column names for the series names or use one column's values for the series names. 

The core issue here is that this update should not have removed an existing feature (allowing metrics to be optional and series names to be dynamic) in exchange for another (interpreting columns as series) when both features are desirable.

**Comment by erikbos** [comment]
Hi @srclosson are you able to comment on the way forward?

**Comment by gilkgit** [comment]
So, not sure if this helps, but working with other tools and also with the default Kusto Explorer visualizations:

for this data:
datatable ( Host: string,Datacenter: string, Time: datetime, Value: int)["H1", "DC1",datetime(2019-10-07),100, "H2", "DC2",datetime(2019-10-07),200]

if I visualize that timeseries with some tools, the series names would be 
1. H1, DC1 
2. H2, DC2

With default Kusto Explorer, I think it would drop the 2nd column and I'll get two series named
1.H1
2.H2
(which will merge series in many cases)

If you want to change that "Default" series naming, you can do stuff like:

datatable ( Host: string,Datacenter: string, Time: datetime)["H1", "DC1",datetime(2019-10-07), "H2", "DC2",datetime(2019-10-07)]
| project Time, Value, SeriesName = strcat('Host: ', Host, ' Datacenter :', Datacenter)

This leaves you with good flexibility.

**Comment by srclosson** [comment]
Let me think, test, and comment before the day is out.

**Comment by mpwillia** [comment]
Thank you for the examples, I understand the change better and I think it should cover all use cases.

However, I'm still a bit lost on where you're defining the aliases such as `$metricname.$value ($tag1|$tag2)` or `$metricname.$value` for example. I know some other data sources having a query builder with an "Alias By" field however I don't have such a field when using the ADX datasource. I only have a field to enter the raw query and a "Format As" option. 

![aliasing-help](https://user-images.githubusercontent.com/7684597/66345896-d5a55000-e905-11e9-85a5-bb4821fe7a51.PNG)

Perhaps I'm just missing something obvious or still not fully understanding? Though I suspect others might have the same lack of knowledge in this regard which might be the root of the misunderstanding here.

**Comment by erikbos** [comment]
@srclosson thanks for the example & explanation, time to give it a try here.. I assume this was based upon the changes of branch MultipleTimeSeries_2019Sep30_Issue09? (@mpwillia that's why you don't have an "Alias By" field yet, it's not yet released)

I'd love to test it but failed to build the plugin, could you help with build instructions or put a test build somewhere? 

(As I could not find a page describing how to build it I used the steps the CI bot used to built your other branch,  but somehow those steps did not led to a working build this time..).

**Comment by mpwillia** [comment]
@erikbos Ah, that makes sense. A bit frustrating that the proposed solution is based off unreleased changes.

@srclosson Do we have an ETA when we could expect at least the "Alias By" field to make it's way to a release? Unfortunately this change seems really rather unfinished without also having the "Alias By" field available. With it though, I think it seems like a fantastic change. In the meantime this will stay a breaking change imo.

**Comment by srclosson** [comment]
I'll ping today too, and see if I can send out a something to test in the meantime. Thanks for the feedback gents. We'll get this closed out soon. @erikbos: I've also resolved another issue you've submitted, so it might make sense to just to a full formal release (ie: 2.1 or something). The other fix allows queries like `show databases` to work.

**Comment by erikbos** [comment]
Excellent, I am happy to test any fixes!

**Comment by srclosson** [comment]
[Download](https://drive.google.com/open?id=1cKp6H8yPKFh4OeeQZxMpZUIDSVF3nIfW) the dist folder here. Just replace your dist folder with this one, and test away. 

You will need to kill the current `grafana-azure-data-explorer-datasource_linux_amd64` process if it's running.

`$ kill $(pidof grafana-azure-data-explorer-datasource_linux_amd64)` should do it.

**Comment by erikbos** [comment]
> [Download](https://drive.google.com/open?id=1cKp6H8yPKFh4OeeQZxMpZUIDSVF3nIfW) the dist folder here. Just replace your dist folder with this one, and test away.
> 
Thanks for making this available, the file does not appear to be publicly readable though.

**Comment by srclosson** [comment]
@erikbos : Added you to the sharing. Weird, google told me that anyone with the link would be able to download. I am confusion. 

With that said, the changes were merged into master just now with docs updated!

**Comment by erikbos** [comment]
> @erikbos : Added you to the sharing. Weird, google told me that anyone with the link would be 
> With that said, the changes were merged into master just now with docs updated!

"Metric Naming & Aliasing" it might make sense to move that section just above the Time Macros section as it feels it belongs more over there.

**Comment by erikbos** [comment]
I don't have time today anymore to test everything.. 

first remark: there is now a difference in behaviour between version v2.0.0 and latest commit:
the column name with all metric names must be now called metric_name,

Example:

let Source = datatable(timestamp:datetime, DeviceModel:string, Count:long)
[
  '2019-10-07T11:01:18', 'iPhone', 32,
  '2019-10-07T11:01:18', 'Android', 432,
  '2019-10-07T11:01:19', 'iPhone', 55,
  '2019-10-07T11:01:19', 'Android', 66,
];
Source 
| project timestamp, DeviceModel, Count
| order by timestamp asc


now requires either:
project timestamp, metric_name=DeviceModel, Count
or
$DeviceModel in 'Alias By'

(v2.0.0 and v1..30 would just adopt DeviceModel)

**Comment by srclosson** [comment]
Hey @erikbos:
Not sure what to suggest. The last little bit could get quite complicated. For example, in a query like

`project timestamp, devicemodel, count, repaired`

It would not be enough to simply say: iPhone or Android. You would need to say something like iPhone.count and Android.repaired... In the simplest case, as in the previous case you mentioned where you literally have a valid, timestamp and metric name, then simply the metric name is enough. 

Is it a big deal that you would need to put "DeviceModel" in the alias by?

I've made a test version locally that will default to something like "iPhone.count", but this is not 1.3 backwards compatible either. But then again, you could not query multiple metrics in 1.3, so this was not an issue. If you want to try the test version, that generates the default dynamically, you certainly can. Otherwise, I would like to close this out based on the fact you can set your own alias.

Let me know what you think.

**Comment by erikbos** [comment]
For me it's mostly about not having to put the column name in 'Alias By' when there is just one string column. That makes it look the most like 1.3 and prevents unnecessary configuration duplication. (Yes, I try to be lazy:))

In case of multiple names just make a combination like you suggestion to have a default?

**Comment by erikbos** [comment]
@srclosson hi, did you manage this close this one?

**Comment by srclosson** [comment]
@erikbos : Sorry Erik, I was travelling last week. What is your default for the column name? metricname? metric? or simply the value column?

**Comment by erikbos** [comment]
I don't think I (or others) have a default... In case of:

let Source = datatable(timestamp:datetime, DeviceModel:string, Count:long)

the default should probably be the string type column? (this is the v1.3 behaviour afaik)

**Comment by srclosson** [comment]
Closing due to inactivity. If incorrect, please feel free to open this again, or submit another issue.

**Comment by erikbos** [comment]
Apologies for the delayed reply, I was out sick and got distracted by another project. I'd love to try the current fixes right now.

I noticed in #9 you mentioned "2.0 re-released and closing this issue." did you mean the current 2.0.0 release would be replaced/updated..? (The plugin installed by grafana-cli does not appear to have the fixes)

After building the plugin I got a few template errors but I somehow don't trust my build (I've noticed in the past the plugin I built the plugin had irregularities I could not explain.)

Are you able to provide: a build/do a release as new version and/or proper build instructions for me?
